internal/cmd/install_test.go

```diff
@@ -0,0 +1,56 @@
+// Copyright 2020 Tetrate
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/tetratelabs/getenvoy/internal/version"
+)
+
+func TestGetEnvoyInstall_VersionValidates(t *testing.T) {
+	o, cleanup := setupTest(t)
+	defer cleanup()
+
+	tests := []struct{ name, version, expectedErr string }{
+		{
+			name:        "version empty",
+			expectedErr: fmt.Sprintf(`invalid <version> argument: "" should look like "%s"`, version.LastKnownEnvoy),
+		},
+		{
+			name:        "version invalid",
+			version:     "a.b.c",
+			expectedErr: fmt.Sprintf(`invalid <version> argument: "a.b.c" should look like "%s"`, version.LastKnownEnvoy),
+		},
+	}
+
+	for _, test := range tests {
+		test := test // pin! see https://github.com/kyoh86/scopelint for why
+
+		t.Run(test.name, func(t *testing.T) {
+			c, stdout, stderr := newApp(o)
+			err := c.Run([]string{"getenvoy", "install", test.version})
+
+			// Verify the command failed with the expected error
+			require.EqualError(t, err, test.expectedErr)
+			// GetEnvoy handles logging of errors, so we expect nothing in stdout or stderr
+			require.Empty(t, stdout)
```

Comment:
ah problem is we had panic bugs everywhere and require helps avoid that #137 I'd prefer to use just require for assertions as panic hunting wastes time.

Reply:
Notice require will immediately fail the test whereas assert will continue with the other assertions until the end. This is better feedback as you won't need to fix error by error but fix them all together.

Reply:
wounds are too fresh, only recently had the luxury of being without soft asserts. I really dislike scanning code up and down to make sure no soft assert isn't relying on state above, and that if that one fails softly another assert isn't going to panic because the reference usually enforced above is no longer good.

life's too hard as it is. TL;DR; I hear ya but can't in good faith go back at this point.

Reply:
ps after a weekend of pondering, I think there were a few things on #137 that led things to be very painful. Even if some are invalidated now, the summary of fixing failed assertions > fixing panics holds.

first, consider a flow like this:

assert not empty versions
assert not empty versions.labels

The overall problem is that the second assert will panic if it isn't a hard assert. There are many forms of this problem especially around if foo, err := stuff. Ex you soft assert err is not nil then if that changes later panic on foo.

1. iirc ginkgo (which luckily is no longer here) would crash strangely on panic. 
2. library upgrades sometimes change structs from chains to chain of pointers
3. panics aren't well formatted, and because they happen later, it is time consuming to figure out what soft assert failed leading to this

The general optimization, in other words, is code for less difficult upgrades, not for "auto-filling tests". Ex I'm aware of the practice of writing a test with a bunch of failed assertions and then try to finish the code. It is indeed slightly faster to do that vs click re-run. However, that's optimizing not just for folks who code like that, but it is in reverse from the concerns above. Basically it makes it quick to add code not well considered and tricky to fix bugs later.

So anyway I really appreciate your feedback and hope that this doesn't come off draconian or anything I generally want to accept changes requested even if they are low-level framework, but I think the "cure" of changing everything to soft asserts isn't something I would do personally as I can't in good faith say this is a nice thing to do for future maintainers including me.